Switch to using Quinoa for UI backend, rather than Node.js + express

holly-cummins opened the pull request:
Resolves #87. 

Here's what I've done: 

### Code changes (the easy part!)

- Replaced node.js express app server with Quarkus app server
- Set up pom.xml for Quinoa
- Moved Angular app to location Quinoa expects
- Converted logic for serving environment variables from node.js to Java
- Removed express dependency from `package.json`

### Testing 

- Tested locally with `quarkus dev`
- Tested locally with `mvn package` and `java -jar target/quarkus-app/quarkus-run.jar`
- Tested locally with `mvn package -Pnative` and `./target/ui-super-heroes-1.0-runner` … the front end starts in 27 ms on my machine, which is pretty awesome
- Tested locally with `docker-compose -f deploy/docker-compose/app.yml up`, although that gives me the old image which is still using node.js, so doesn’t tell me much

### Deployment and automation

- Removed custom docker files and replaced them with standard Quarkus ones
- Added  Quarkus Kubernetes extensions into the UI pom.xml
- Removed the special casing for UI projects in generate-k8s-resources
- Removed the special UI build steps from the GitHub workflows and added the ui project to the matrix builds. (The same matrix seems to be repeated three times, so I wonder if it should be common?)
- Updated `app.openshift.io/runtime: nodejs` to specify quarkus in all the Kubernetes .ymls, but I wasn’t sure how to test this

### Documentation 

- Updated readme (including architecture diagrams)
- Updated the azure instructions for running the UI app locally
- Updated automation documentation to remove special-casing of UI app. I hand-updated a diagram with a workflow screencap but we might want to regenerate it when this is merged and the workflows are running.

In both the k8s script and the automation docs we list out each project individually, to be able to exclude the UI project. That’s not needed anymore, so we may want to do some `*`-ing or “all” in the text, but I didn’t make those changes. We may also spot some other simplifications that come from having a homeogenous stack across all projects.

I’m not totally sure I have the instructions for Azure Static Web Apps right, but it’s hard to test. We’re still generating the same static app as we were before. Before it would ask a node.js server for its config, now it asks a Quarkus server for its config, so I’m not sure how either model exactly works with a static deployment.

edeandrea commented:
Thank you @holly-cummins ! I'll take a closer look tomorrow or Friday, but first thing I noticed is that you shouldn't manually change anything in a folder named `deploy`, whether at the root of the project or inside one of the application folders. Everything in those folders are generated as part of the CI/CD process.

holly-cummins commented:
Ah, thanks, @edeandrea. It seemed wrong to leave the wrong values in source control. :) 

I assume CI/CD will just overwrite what's there with new content (which would be the same as what I generated), so my changes wouldn't do any harm?

edeandrea commented:
> I assume CI/CD will just overwrite what's there with new content (which would be the same as what I generated), so my changes wouldn't do any harm?

Yeah the CI process will essentially do an `rm -rf **/deploy/**`, regenerate everything, and then commit it back into source control.

Just for future reference - no need to ever touch anything inside any directory named `deploy` :)

I'll take a closer look at this tomorrow/friday and comment on it.

edeandrea commented:
We could. I tried to make it obvious by having a README file in each directory saying not to manually modify anything in there as it'll be overwritten by the CI process.

Apparently I failed there too :)

edeandrea commented:
Yeah you'd never want to run the `build-push-container-images` workflow on a fork because of that. It requires the token. Really only the simple-build-test (to verify the app is ok), but we don't have a good way of verifying that all the k8s yamls/etc all "look" right unfortunately.

edeandrea commented on ui-super-heroes/pom.xml:
```diff
@@ -6,4 +6,137 @@
+    <dependency>
+      <groupId>io.quarkiverse.quinoa</groupId>
+      <artifactId>quarkus-quinoa</artifactId>
+      <version>2.0.3</version>
```
Maybe set the version as a property in the `<properties>` section? And maybe version `2.0.4` (the current latest)?

edeandrea commented on ui-super-heroes/pom.xml:
```diff
@@ -6,4 +6,137 @@
+    <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
+    <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
+    <quarkus.platform.version>3.0.3.Final</quarkus.platform.version>
+    <skipITs>true</skipITs>
```
Should remove this to be consistent with the rest of the apps (we ALWAYS run integration tests).

edeandrea commented on ui-super-heroes/pom.xml:
```diff
@@ -6,4 +6,137 @@
+    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+    <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
+    <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
+    <quarkus.platform.version>3.0.3.Final</quarkus.platform.version>
```
Should be `3.1.0.Final` to be consistent with the rest of the apps.

edeandrea replied:
Should be `3.1.1.Final` to be consistent with the rest of the apps.

edeandrea commented on ui-super-heroes/src/main/docker/Dockerfile.jvm:
```diff
@@ -0,0 +1,93 @@
+#
+# Then, build the image with:
+#
+# docker build -f src/main/docker/Dockerfile.jvm -t quarkus/quinoa-app-jvm .
```
`quarkus/quinoa-app-jvm` _should_ be `quarkus/ui-super-heroes-jvm17` (to be consistent with all the other apps)

edeandrea commented on ui-super-heroes/src/main/docker/Dockerfile.jvm:
```diff
@@ -0,0 +1,93 @@
+#
+# Then run the container using:
+#
+# docker run -i --rm -p 8080:8080 quarkus/quinoa-app-jvm
```
`quarkus/quinoa-app-jvm` _should_ be `quarkus/ui-super-heroes-jvm17` (to be consistent with all the other apps)

edeandrea commented on ui-super-heroes/src/main/docker/Dockerfile.jvm:
```diff
@@ -0,0 +1,93 @@
+#
+# Then run the container using :
+#
+# docker run -i --rm -p 8080:8080 quarkus/quinoa-app-jvm
```
`quarkus/quinoa-app-jvm` _should_ be `quarkus/ui-super-heroes-jvm17` (to be consistent with all the other apps)

edeandrea commented on ui-super-heroes/src/main/docker/Dockerfile.legacy-jar:
```diff
@@ -0,0 +1,90 @@
+#
+# Then, build the image with:
+#
+# docker build -f src/main/docker/Dockerfile.legacy-jar -t quarkus/quinoa-app-legacy-jar .
```
`quarkus/quinoa-app-jvm` _should_ be `quarkus/ui-super-heroes-jvm17` (to be consistent with all the other apps)

edeandrea commented on ui-super-heroes/src/main/docker/Dockerfile.legacy-jar:
```diff
@@ -0,0 +1,90 @@
+#
+# Then run the container using:
+#
+# docker run -i --rm -p 8080:8080 quarkus/quinoa-app-legacy-jar
```
`quarkus/quinoa-app-jvm` _should_ be `quarkus/ui-super-heroes-jvm17` (to be consistent with all the other apps)

edeandrea commented on ui-super-heroes/src/main/docker/Dockerfile.legacy-jar:
```diff
@@ -0,0 +1,90 @@
+#
+# Then run the container using :
+#
+# docker run -i --rm -p 8080:8080 quarkus/quinoa-app-legacy-jar
```
`quarkus/quinoa-app-jvm` _should_ be `quarkus/ui-super-heroes-jvm17` (to be consistent with all the other apps)

edeandrea commented on ui-super-heroes/src/main/docker/Dockerfile.legacy-jar:
```diff
@@ -0,0 +1,90 @@
+#   accessed directly. (example: "foo.example.com,bar.example.com")
+#
+###
+FROM registry.access.redhat.com/ubi8/openjdk-17:1.14
```
Should be `FROM registry.access.redhat.com/ubi9/openjdk-17:1.14` to be consistent with the rest of the apps.

edeandrea commented on ui-super-heroes/src/main/kubernetes/openshift.yml:
This file probably shouldn't even exist. Everything in here will be generated by the `quarkus-openshift` extension.

edeandrea reviewed (changes requested):
I spent some time going through all of this and have made some comments in-line.

I also think that the docker compose stuff is missing. See `src/main/docker-compose` from some of the other apps as well as the `scripts/generate-docker-compose-resources.sh`. There probably need to be modifications in there as well.

Also you should check the `scripts/deploy-to-azure-containerapps.sh` script and verify nothing needs to change there.

edeandrea reviewed (changes requested):
I spent some time going through all of this and have made some comments in-line.

I also think that the docker compose stuff is missing. See `src/main/docker-compose` from some of the other apps as well as the `scripts/generate-docker-compose-resources.sh`. There probably need to be modifications in there as well.

Also you should check the `scripts/deploy-to-azure-containerapps.sh` script and verify nothing needs to change there.

holly-cummins commented on .github/workflows/build-push-ui-images.yml:
Deleting it from one branch shouldn't affect the other branches, so I think it's safe to remove it in `main`. It will still appear in the sidebar as a browsable workflow.

holly-cummins replied:
(In fact, apparently it's quite hard to get rid of workflows altogether. I was talking to Guillaume about tidying up some that I'd created for Quarkus, and the conclusion was that it was more trouble that it was worth.)

edeandrea replied:
But if i want to trigger the workflow manually from the browser it has to be on the `main` branch I believe, which is what we want with this particular workflow.

holly-cummins replied:
Hmm, yes, [the docs](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow#running-a-workflow-using-github-cli) seem to confirm you can only do `workflow_dispatch` if the workflow exists on the default branch. 

So then I guess it's a tradeoff - we could edit the file to *only* have the `workflow_dispatch` trigger on `main`, but it's potentially confusing clutter ... and we're trying to maintain a workflow on a branch where that workflow doesn't make sense, so it would be easy to get it wrong. I guess it's a tradeoff, and a question of how often we trigger that workflow manually, and how often we use the `workflow_call` trigger. (I'd assume we

edeandrea replied:
We actually use the workflow quite a lot. We've had a long-standing issue where for some reason, the image on quay stops working. It fails on openshift due to some root permission issue. I've spent countless weeks trying to figure out why, but if I push the image today, everything is fine, I don't touch the image for weeks, but then all of a sudden starts failing. Rebuilding & repushing the image (from the same `Dockerfile` and same source tree) fixes it.

Even though we most likely won't need that capability on the `main` branch anymore, we still need it on the `rhbq-2.7` and `rhbq-2.13` branches though.

Just 2 weeks ago at Red Hat summit we had the problem again, and that's when I created this quick way just to fix the issue for this one node.js image.

Over time we can certainly eliminated it (maybe after the next 2 RHBQ releases), but for now it really does need to stay.

holly-cummins replied:
Ah, that makes sense. Ok, reinstated!

holly-cummins commented on .github/workflows/build-push-container-images.yml:
```diff
@@ -350,77 +293,6 @@ jobs:
-        working-directory: ui-super-heroes
-        run: "docker push -a ${{ env.IMAGE_BASE_NAME }}/ui-super-heroes"
-
   create-app-multiarch-manifests:
```
I did indeed, good spot. 

Although my error is more evidence that maybe one big matrix is the way to go. :)

edeandrea replied:
maybe now after the restructuring that does make more sense, although it is broken up because we want to build all the images of the matrix before pushing a single one, and then only create the multi-arch manifest only once all the images are pushed, so it may still need to remain as separate jobs.

holly-cummins commented on ui-super-heroes/README.md:
```diff
@@ -23,34 +23,39 @@ Environment variables can be injected into the build using the [ngx-env](https:/
 3. Set the `API_BASE_URL` environment variable with the appropriate [Fights Service](../rest-fights) hostname and port.
    > By default, the [`rest-fights`](../rest-fights) service runs on port `8082`, so setting `API_BASE_URL=http://localhost:8082` will do.
-4. Start the service using the command `npm start`.
+4. Start the service using the command `quarkus dev`.
```
I think that's a good idea, but maybe for another work item? I was trying to keep this one as a like-for-like swap, so that the new service behaved the same as the old one.

edeandrea replied:
I thought the other app did default to localhost:8082? Or maybe it didn't? Honestly I don't remember :)

holly-cummins replied:
Ah, I see what you mean. You're right, it should run correctly with no config. The readme even says so: 

```
Currently, the `env.js` will expose just the `API_BASE_URL` that's set at runtime. This will control the base URL to connect to the [fights](../rest-fights) service. The default if unset is http://localhost:8082.
```

That's already implemented, with this, in `EnvResource.java`

```
    @ConfigProperty(name = "api.base.url", defaultValue = "http://localhost:8082")
    String url;
 ```   
 
 The readme should probably be updated to explain that it works by normal Quarkus config so you can use `api.base.url` or an env variable of `API_BASE_URL`.

holly-cummins commented on ui-super-heroes/src/main/kubernetes/kubernetes.yml:
Ah, that makes sense. I'd assumed I was safe changing nothing in that file, but you're right that we have the kubernetes plugin which means we can save a whole bunch of yaml. 
I've renamed to `common.yml` to follow the pattern of the others, and deleted everything but the `Ingress`. Let me know if that seems right.

holly-cummins replied:
Ah, nevermind, I've just seen the comment on the following file, so will un-delete!

edeandrea replied:
Yeah no we don't want `common.yml` since we don't want the `Ingress` in the openshift generation. The files need to remain as is.

edeandrea replied:
Yes - its in [another comment](https://github.com/quarkusio/quarkus-super-heroes/pull/297#discussion_r1222022308)

holly-cummins replied:
Oh wait, I think I meant `knative.yml`!

edeandrea replied:
Thats in yet [another comment](https://github.com/quarkusio/quarkus-super-heroes/pull/297#discussion_r1222021078) :)

holly-cummins commented on ui-super-heroes/src/main/resources/application.properties:
Wow, that's a lot of config. It makes me think we're missing some defaults. But I've added it all in. 

Do you know what the difference is between `app=` and `application=` in the otel attributes?

edeandrea replied:
Its consistent with the labels we use in the kubernetes descriptors. In the case of this app, though, their value is probably the same because its a standalone app with nothing else. In the case of something like the villains app, `app=rest-villains` and `application=villains-service`.

As far as the rest of the config goes, we tried to introduce profiles to make it easier for users to deploy to various k8s variants without having to remember all the params they need to set. See https://github.com/quarkusio/quarkus-super-heroes/tree/main/rest-villains#deploying-directly-via-kubernetes-extensions for an example.

The `README.md` file in `ui-super-heroes` should probably also contain a lot of the same information as well now that there are multiple variants.

holly-cummins replied:
If we did 

```
%openshift.quarkus.config.profile.parent=kubernetes
```

instead of 

```
%openshift.quarkus.config.profile.parent=prod
```

wouldn't we be able to drop a whole bunch of the copy-and-pasted config for the kubernetes-variants profiles?

And I feel like there should be a default for stuff like this:

```
quarkus.kubernetes.labels.app=${quarkus.application.name}
```

edeandrea replied:
I'm not sure if the openshift extension will inherit things from the `quarkus.kubernetes` namespace, because there is the `quarkus.openshift` config namespace, and the documentation doesn't say anything about any kind of inheritance.

So, maybe, but untested.

holly-cummins replied:
Ah, that's a good point about the namespace.

edeandrea commented:
One other thing I noticed that I forgot to mention in the review - you'll need to modify the `simple-build-test` workflow and add the `ui-super-heroes` app to both of the matrix jobs.

holly-cummins commented on ui-super-heroes/src/main/resources/application.properties:
```diff
@@ -0,0 +1,5 @@
+quarkus.quinoa.package-manager-install=true
+quarkus.quinoa.package-manager-install.node-version=16.16.0
```
I've just remembered, looking at the broken build, why we're on node 16: https://stackoverflow.com/questions/69692842/error-message-error0308010cdigital-envelope-routinesunsupported

It seems like various bits of the ecosystem need to catch up with node 18. Maybe, when we do #290 we can move up node versions, but for now, I think we should stay on 16.

edeandrea replied:
That's interesting because its currently running on node 18...

https://github.com/quarkusio/quarkus-super-heroes/blob/2f99db07cfe69aa4313a9337fb2c76761d66850f/ui-super-heroes/Dockerfile#L23

holly-cummins replied:
Ah, it's this line, a little bit further down in the same dockerfile, that makes it work 

```
# NODE_OPTIONS=--openssl-legacy-provider is a hack, but until the UI has been completely re-written, this will have to do
ENV HTTP_PORT=8080 \
    NODE_ENV=production \
    NODE_OPTIONS=--openssl-legacy-provider
```

So we could do the same setting of `NODE_OPTIONS`, but the effect is similar to downgrading to 16; in either case we're running with the less-secure SSL, so I'm not sure there's much in it.

edeandrea replied:
Yeah I guess it doesn't matter. We should probably put that in the node options anyways just to keep it the same.

holly-cummins replied:
My worry with putting it in the node options is that then we have to remember to take it out of the node options when we do #290 ... (We also have to remember to update Node when we do #290, but that seems maybe easier to remember?)

edeandrea replied:
I guess you're right

holly-cummins commented:
> I spent some time going through all of this and have made some comments in-line.
> 
> I also think that the docker compose stuff is missing. See `src/main/docker-compose` from some of the other apps as well as the `scripts/generate-docker-compose-resources.sh`. There probably need to be modifications in there as well.

Done! 

For another WI, do you think it's possible to reduce the volume of docker compose files? I was a bit perplexed by the relationship between `rest-fights/deploy/docker-compose/native.yml` and `rest-fights/src/main/docker-compose/native.yml`, for example. One seems to be generated and the other is hand-maintained? For the UI project they're basically identical, but for others there's overlap, but also different content. I couldn't figure out the intention from a search of the readmes. (And a readme probably isn't the right place for that kind of information, because people might not be able to find it mixed in with all the other content, and when they need is when they're looking at the file, trying to understand if it's ok to update or if it's generated.)

> Also you should check the `scripts/deploy-to-azure-containerapps.sh` script and verify nothing needs to change there.

Good point, and done! I think because the externals of the UI container are exactly the same, no changes are needed.

edeandrea commented:
For the docker compose stuff, yes src/main are hand maintained "snippets" whereas in the deploy directory is the "real" docker compose file. It's done this way so that we can have the root `/deploy/docker-compose` with an uber compose file that had everything from all the apps in it.

It was done that way to try and mimic the kubernetes extension. If we did away with that we'd have to hand craft 2 sets of compose files - 1 individually for each app and 1 uber compose file, both for jvm (and perhaps multiple jvm versions) and native.

holly-cummins commented:
> For the docker compose stuff, yes src/main are hand maintained "snippets" whereas in the deploy directory is the "real" docker compose file. It's done this way so that we can have the root `/deploy/docker-compose` with an uber compose file that had everything from all the apps in it.
> 
> It was done that way to try and mimic the kubernetes extension. If we did away with that we'd have to hand craft 2 sets of compose files - 1 individually for each app and 1 uber compose file, both for jvm (and perhaps multiple jvm versions) and native.

Got it, thanks. I've moved the discussion to https://github.com/quarkusio/quarkus-super-heroes/issues/299.

holly-cummins commented:
Ok, I *think* I've got all the comments. Thanks for spotting the issues, @edeandrea!

edeandrea commented:
Assuming that means you think it's ready for another pass through? :)

Of so, I'll try to get to it later today

holly-cummins commented:
> Assuming that means you think it's ready for another pass through? :)
> 
> Of so, I'll try to get to it later today

Yes please, thank you!

edeandrea commented:
@someth2say FYI once this is merged it's going to have an effect on #291. Should actually make the work you are doing there easier I would think, although you may get some merge conflicts. Please be aware.

edeandrea commented on deploy/docker-compose/native.yml:
No need to modify this file since it lives in the `/deploy` directory.

edeandrea commented on ui-super-heroes/README.md:
Thinking that we should make this README be similar to the style of the READMEs of the other projects? We don't have to have the table of exposed endpoints. What do you think?

I'm thinking documentation like `Running the Application`, `Running Locally via Docker Compose`, `Deploying to Kubernetes`, `Using pre-built images`, & `Deploying directly via Kubernetes Extensions`?

holly-cummins replied:
To be honest, enthusiasm is a bit low. There's a lot of content in this readme which is unique to the UI, so we wouldn't want to stick so closely to the structure of the others that we lose that content. I'm also not sure we need instructions on things like deploying to kubernetes if it's fairly similar across all applications; it could be repetitive boilerplate that masked the 'interesting' information, making the docs hard to use. 

Certainly, I think that kind of rewrite feels outside the scope of this PR; the readme has been updated to reflect the new application architecture, including things like updated run instructions and architecture diagrams and pointers on running natively. Doing a more general structural review is maybe a separate work item?

edeandrea replied:
> To be honest, enthusiasm is a bit low

Everyone loves writing documentation, right? :D

> Doing a more general structural review is maybe a separate work item?

I guess we can see what it looks like once its merged, but I do think having the sections talking about the profiles, how to deploy to kube using the maven commands/quarkus cli is valuable to have, rather than having the user having to go back and look at another project's docs to remember the specific command to use.

holly-cummins replied:
I agree that having information close to where it's needed is valuable, so it's a balance between that and duplicating information so widely that it's a maintenance burden because every update needs to be mirrored across many readmes.

In either case, I consider writing documentation about the quarkus kubernetes extensions and cli to be outside the scope of this work item. :)

edeandrea replied:
For me its about making sure that the documentation is usable. This is a big structural change for this app, which means that the documentation probably needs a structural change as well.

This change completely changes how the app gets run, how it gets deployed to various platforms (i.e. kube, knative, openshift, etc) and what "versions" of the app are available.

We can see what it looks like once its merged and then update it accordingly after I suppose.

holly-cummins replied:
There's an argument, this does not change how the app gets run. I was very careful to make sure the externals stayed exactly the same. Endpoints stayed the same, it gets configured by environment variable in exactly the same way, the externals of the container are exactly the same ... 

Actually, it might be more accurate to say that sometime last week this PR didn't change how the app was run, but now we've done a whole bunch of extra work which does change how the app gets run. Last week we treated the application as a black box which was packaged in a docker container, but now we've inserted a whole bunch of extra configuration options to allow us to configure how it gets deployed to kubernetes, knative, etc. However, I don't think any of that extra work makes the current documentation incorrect. At the moment it says: 

> ## Deploying to Kubernetes
> Pre-built images for this application can be found at [`quay.io/quarkus-super-heroes/ui-super-heroes`](https://quay.io/repository/quarkus-super-heroes/ui-super-heroes?tab=tags).
> Deployment descriptors for this image are provided in the [`deploy/k8s`](deploy/k8s) directory. There is one for [OpenShift](https://www.openshift.com) ([`app-openshift.yml`](deploy/k8s/app-openshift.yml)), [Minikube](https://minikube.sigs.k8s.io) ([`app-minikube.yml`](deploy/k8s/app-minikube.yml)), and [Kubernetes](https://www.kubernetes.io) ([`app-kubernetes.yml`](deploy/k8s/app-kubernetes.yml)).
 
 There may be some extra documentation about the quarkus kubernetes extension we can copy and paste in, but that just feels wildly outside the scope of this work item, whose original goal was to show that Node.js could be replaced with Quarkus for a relatively modest effort.

edeandrea replied:
> this does not change how the app gets run

If you're just talking about how to `docker run` the app from the container image that is produced, then you are completely 100% accurate. The app is run exactly the same way it was before. From a black box perspective it is still run the same way.

> but now we've done a whole bunch of extra work which does change how the app gets run.

I don't think that's entirely accurate, though. We've changed how the container image itself gets produced, along with the automation which produces all the various configs/yamls/docker compose/etc, which is vastly different than how it was being done before when it was just a Node.js project using express.js.

Which also brings along new ways of actually deploying/running the app on kubernetes. There was no way beforehand to run a node.js command and "boom" its now running on kube, which we now have with maven/Quarkus CLI.

> whose original goal was to show that Node.js could be replaced with Quarkus for a relatively modest effort

That might be true, but because this is a sample app, we also need to make sure that users know and understand all the various options and how to do things. For example, if they just wanted to deploy this one app to openshift in native mode, how would they do that?

All that being said, we can leave the documentation as-is in this PR and re-evaluate later if we think it needs to be enhanced.

edeandrea commented on ui-super-heroes/deploy/docker-compose/native.yml:
This change isn't really needed because it'll be generated by the CI process.

edeandrea commented on ui-super-heroes/deploy/docker-compose/java17.yml:
This change isn't really needed because it'll be generated by the CI process.

edeandrea commented on ui-super-heroes/pom.xml:
```diff
@@ -6,4 +6,137 @@
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-minikube</artifactId>
+    </dependency>
+    <dependency>
```
This comment still stands as unresolved.

edeandrea commented on ui-super-heroes/src/main/docker-compose/java17.yml:
The `ui-super-heroes` service name should be `ui-super-heroes-java17`

edeandrea commented on ui-super-heroes/src/main/java/io/quarkus/workshop/superheroes/ui/EnvResource.java:
Shouldn't there be tests for this class (both unit & IT)?

edeandrea commented on ui-super-heroes/src/main/java/io/quarkus/workshop/superheroes/ui/EnvResource.java:
I think we should also have a test class that verifies (as basic as possible) that the UI actually loads?

edeandrea commented on ui-super-heroes/src/main/kubernetes/kubernetes.yml:
```diff
@@ -1,82 +1,3 @@
-    app: ui-super-heroes
-    application: super-heroes
-  type: ClusterIP
----
```
The file should start with the `---`

edeandrea reviewed (changes requested):
We're close! See comments in-line.

edeandrea commented:
> I don't think there's any point in getting too focussed on specific versions, is there

Just that the simple-build-test workflow runs against this PR. Not that I expect any failures, but the version update _could_ potentially incorporate a problem which then would cause the main CI to break when this gets merged to `main`.

But you're right, probably not a big deal in this case.

holly-cummins commented:
> > I don't think there's any point in getting too focussed on specific versions, is there
> 
> Just that the simple-build-test workflow runs against this PR. Not that I expect any failures, but the version update _could_ potentially incorporate a problem which then would cause the main CI to break when this gets merged to `main`.
> 
> But you're right, probably not a big deal in this case.

I think if the version of Quarkus in one microservice being a week out of date caused that kind of problem, we would need to go back to the Quarkus team and have a chat about their release testing. :)  

Although, actually, I don't think I entirely see the source of risk. If we think the `simple-build-test workflow` that we run against PRs isn't comprehensive enough to catch some problems, and we think the move from Quarkus 3.0.3 to 3.1.1 is a risk, wouldn't that risk still be there if I updated before merging? Because I'd be merging with Quarkus 3.1.1, but we wouldn't be confident of catching problems with that version? Or would we be assuming I'd be running more comprehensive tests locally with 3.1.1 than the PR workflow runs? 

I think I'm missing something about the process and the concern.

edeandrea commented:
> I think I'm missing something about the process and the concern.

This particular app has very little (meaning close to 0) tests, whereas all the other apps have quite a big test suite, meaning the simple build test workflow is definitely trustworthy for all the other apps. In the year and a half I've been maintaining this, I've never had an issue where something was green in the PR and then things were broken once merged (in the other apps - I've had it happen with the UI app, which is why I've tried to not touch it with a 10-foot pole :) ). Hopefully #290 will help there.

edeandrea commented on ui-super-heroes/src/main/resources/application.properties:
```diff
@@ -0,0 +1,78 @@
+quarkus.knative.min-scale=1
+
+quarkus.openshift.env.vars.CALCULATE_API_BASE_URL=true
+quarkus.knative.env.vars.CALCULATE_API_BASE_URL=true
```
Thats actually not a bad idea, and I like it, but probably out of scope for this PR.

holly-cummins replied:
Agreed, definitely out of scope. :)

holly-cummins commented:
> > I think I'm missing something about the process and the concern.
> 
> This particular app has very little (meaning close to 0) tests, whereas all the other apps have quite a big test suite, meaning the simple build test workflow is definitely trustworthy for all the other apps. In the year and a half I've been maintaining this, I've never had an issue where something was green in the PR and then things were broken once merged (in the other apps - I've had it happen with the UI app, which is why I've tried to not touch it with a 10-foot pole :) ). Hopefully #290 will help there.

That makes sense. I'm just not convinced that there's much correlation between someone being a human-dependabot and updating a pom.xml before merging and them doing manual testing. :) 

As you say, I think the solution is to improve the testing, not to try and do all version updates manually. Even if I updated to Quarkus 3.1.1 before merging, you know 3.1.2 would be along shortly, and then we have the exact same issue - "can we trust a version update if there's not someone to do manual testing of it before merging?"

holly-cummins commented:
Ok, I’ve done the next round of changes, *except* for the following (as discussed above): 

- Analyse all the readmes and scripts to identify files which are generated, and avoid changing them (#299 should help with that in future)
- Monitor dependabot changes in the rest of the codebase and manually mirror them to this PR while the change is in review
- Restructure readme to include more instructions about how the quarkus kubernetes extension works

You’re quite right, @edeandrea, there should have been tests for the Java part. As there weren’t any for the old express server, I had kept parity, but we should aim higher.  

I’ve also added the [Quinoa incantation](https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/index.html#_testing) to run the Javascript tests, should we ever add any (#290). The javascript tests were so missing that `npm run test` wasn’t even a no op, it actually failed. So I’ve fixed that, although it ended up involving adding a skeleton test and loads of fixes to make the test framework work at all. The application didn’t even have any of the skeletons for testing, and finding an example of an angular test setup that could worked with Angular 12 was a bit arduous. (And, ultimately pointless, since we know we need to move away from Angular 12.) Retrofitting modern examples, of course, does not work. In the end, I got most of the way there but Jasmine insisted there were no test files. I spent about three hours debugging that but it really doesn’t seem like a good use of time, so in the end I just bypassed the actual Javascript test execution, but left the scaffolding I’d got and the Quinoa executor in.

I also spotted an incorrect package in the Java code, so I’ve fixed that.

edeandrea commented:
Thank you very much @holly-cummins! I'll merge once CI has run.

holly-cummins commented:
Thanks, @edeandrea!

edeandrea commented:
There were a few things I missed too in my reviews that I realized when I checked out the branch. I just committed them myself to this PR.
